test/distributed/_sharded_tensor/ops/test_linear.py

```diff
@@ -84,14 +88,28 @@ def _run_sharded_linear(self, spec, input_size, linear_size, sharded_dim):
             local_linear.weight, sharded_dim, TEST_GPU_NUM, spec, self.rank
         )
         local_grad_narrowed = local_grad.narrow(sharded_dim, start_pos, chunk_size)
-        local_weight_narrowed = local_linear.weight.narrow(
-            sharded_dim, start_pos, chunk_size
-        )
 
         # Test backward gradient calculation.
         self.assertEqual(sharded_linear.bias.grad, local_linear.bias.grad)
         self.assertEqual(sharded_weight.grad, local_grad_narrowed)
 
+        # Test optimizer.
+        previous = local_linear.weight.clone().detach()
+        optim = torch.optim.SGD([local_linear.weight], lr=0.1)
+        optim.step()
+        self.assertNotEqual(previous, local_linear.weight)
+        previous_sharded = sharded_weight.clone().detach()
+        sharded_optim = ShardedOptimizer(dict(named_params_with_sharded_tensor(sharded_linear)), torch.optim.SGD, lr=0.1)
+        sharded_optim.step()
+        sharded_weight = sharded_linear.weight.local_shards()[0].tensor
+        local_weight_narrowed = local_linear.weight.narrow(
+            sharded_dim, start_pos, chunk_size
+        )
+        self.assertEqual(sharded_weight.size(), local_weight_narrowed.size())
+        with torch.no_grad():
```

Comment:
Why do we need `torch.no_grad` here?

Reply:
It complains if not added. Because here `previous_sharded` is from detach() which does not come with gradient. And looks like the result from narrow also does not have gradient.